### PR TITLE
Target Android 14

### PR DESCRIPTION
**Pull request by AaronChen0**
This is an attempt to update shadowsocks-android to target Android 14 and its dependencies.

Please read [this](https://developer.android.com/build/releases/past-releases/agp-8-0-0-release-notes#default-changes) for changes in AGP8. Some build option default values change.

The build succeeded and the debug apk works in both my android 14 and android 12 phones.
You can try a debug build apk [here](https://app.circleci.com/pipelines/circleci/YBg2SAd5kz9ouc3TKNZvz3/KRnxuTDpRx85DnYaX86Nc6/57/workflows/81f735a5-1082-46ee-b483-5cc0eaa8b581/jobs/57/artifacts).
The debug build log can be found [here](https://circleci.com/api/v1.1/project/circleci/YBg2SAd5kz9ouc3TKNZvz3/KRnxuTDpRx85DnYaX86Nc6/57/output/107/0?file=true&allocation-id=66cc40d867a4fc7077e4f5cb-0-build%2FABCDEFGH).
 
Main updates:
* Bump shadowsocks-rust from `1.15.3` to `1.20.4`
* Bump gradle from `7.6` to `8.10`
* Bump Android Gradle plugin from `7.4.2` to `8.5.2`
* Bump Kotlin from `1.8.10` to `2.0.10`
* Migrate from kapt to KSP. Ref: https://developer.android.com/build/migrate-to-ksp
* Add foreground service type to manifest. Ref: https://developer.android.com/about/versions/14/changes/fgs-types-required

### CODE REVIEW

**Comment by AaronChen0** [comment]
In a release branch [here](https://github.com/AaronChen0/shadowsocks-android/tree/android14_release), the build will fail with the following R8 missing class errors:
```
WARNING: /home/circleci/code/mobile/build/intermediates/merged_java_res/release/mergeReleaseJavaResource/base.jar: R8: Unexpected reference to missing service class: META-INF/services/java.net.spi.InetAddressResolverProvider.
WARNING: /home/circleci/code/mobile/build/intermediates/merged_java_res/release/mergeReleaseJavaResource/base.jar: R8: Unexpected reference to missing service implementation class in META-INF/services/java.net.spi.InetAddressResolverProvider: org.xbill.DNS.spi.DnsjavaInetAddressResolverProvider.
WARNING: /home/circleci/code/mobile/build/intermediates/merged_java_res/release/mergeReleaseJavaResource/base.jar: R8: Unexpected reference to missing service class: META-INF/services/sun.net.spi.nameservice.NameServiceDescriptor.
ERROR: Missing classes detected while running R8. Please add the missing classes or apply additional keep rules that are generated in /home/circleci/code/mobile/build/outputs/mapping/release/missing_rules.txt.
ERROR: R8: Missing class lombok.Generated (referenced from: org.slf4j.Logger org.xbill.DNS.Compression.log and 3 other contexts)
Missing class org.slf4j.impl.StaticLoggerBinder (referenced from: void org.slf4j.LoggerFactory.bind() and 3 other contexts)
Missing class org.xbill.DNS.spi.DnsjavaInetAddressResolverProvider (referenced from: java.util.List kotlinx.coroutines.internal.FastServiceLoader.load(java.lang.Class, java.lang.ClassLoader))
Missing class sun.net.spi.nameservice.NameServiceDescriptor (referenced from: org.xbill.DNS.spi.DNSJavaNameServiceDescriptor)
```
Adding the following to `proguard-rules.pro` fixes the R8 errors and the apk works normally.
```
-dontwarn java.net.spi.InetAddressResolverProvider
-dontwarn org.slf4j.impl.StaticLoggerBinder
-dontwarn org.xbill.DNS.spi.DnsjavaInetAddressResolverProvider
-dontwarn sun.net.spi.nameservice.NameServiceDescriptor
-dontwarn lombok.Generated
```
I am not familiar with R8 proguard rules, should I just add `dontwarn` if the app works normally in my use?
Check the builds [here](https://app.circleci.com/pipelines/circleci/YBg2SAd5kz9ouc3TKNZvz3/KRnxuTDpRx85DnYaX86Nc6?branch=android14_release).

**Comment by Mygod** [comment]
Yes add these rules.

```
-dontwarn lombok.Generated
-dontwarn org.slf4j.impl.StaticLoggerBinder
-dontwarn org.xbill.DNS.spi.DnsjavaInetAddressResolverProvider
-dontwarn sun.net.spi.nameservice.NameServiceDescriptor
```

File: core/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -43,24 +46,36 @@
...
             android:label="@string/app_name"
             android:permission="android.permission.BIND_VPN_SERVICE"
+            android:foregroundServiceType="specialUse"
             android:exported="false">
             <intent-filter>
                 <action android:name="android.net.VpnService"/>
             </intent-filter>
+            <property
+                android:name="android.app.PROPERTY_SPECIAL_USE_FGS_SUBTYPE"
+                android:value="Vpn" />
```
Comment (Mygod) [inline core/src/main/AndroidManifest.xml:56]: https://developer.android.com/about/versions/14/changes/fgs-types-required#system-exempted

File: mobile/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -99,11 +99,15 @@
...
                  android:icon="@drawable/ic_service_active"
+                 android:foregroundServiceType="specialUse"
                  android:permission="android.permission.BIND_QUICK_SETTINGS_TILE"
                  tools:targetApi="n">
             <intent-filter>
                 <action android:name="android.service.quicksettings.action.QS_TILE" />
             </intent-filter>
+            <property
+                android:name="android.app.PROPERTY_SPECIAL_USE_FGS_SUBTYPE"
+                android:value="tile" />
```
Comment (Mygod) [inline mobile/src/main/AndroidManifest.xml:110]: This is not necessary.

**Review by Mygod — CHANGES_REQUESTED** [review]

**Comment by AaronChen0** [comment]
Build successfully and work in my phones. Please review.

**Comment by Mygod** [comment]
Thanks! And be patient while I download rust. :)

**Review by Mygod — APPROVED** [review]
LGTM

**Pull request merged**